# Stage A Tag-along Project
**ID:** 14d441dfa201f000

In [1]:
# importing the needed libraries
import numpy as np
import pandas as pd

### Source of Data
The dataset is provided by the Food and Agriculture Organization of the United Nations and can be gotten through the HamoyeHQ github repo shown bellow.
https://github.com/HamoyeHQ/HDSC-Introduction-to-Python-for-machine-learning 

In [2]:
# loading data into pandas DataFrame
food_data = pd.read_csv('foodbalancesheets_e_africa_noflag.csv',encoding='latin-1')
food_data.head()

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

### Question 11.
What is the total sum of Animal Fat produced in 2014 and 2017 respectively?

In [3]:
# Solution
# First we filter out the dataset using the 'Item' feature to get the data for only "Animal fats"
# Then we sellect desired year columns 'Y2014' and 'Y2017', them call sum() on the result as follows
animal_fat_2014_2017 = food_data[food_data.Item == 'Animal fats'][['Y2014', 'Y2017']].sum()

#sellecting the values for each year
animal_fat_2014 = animal_fat_2014_2017.loc['Y2014']
animal_fat_2017 = animal_fat_2014_2017.loc['Y2017']

#printing the result
print('Total Animal Fat for 2014 and 2017')
print('==='*11)
print(f'Y2014 Animal fat: {animal_fat_2014}')
print(f'Y2017 Animal fat: {animal_fat_2017}')

Total Animal Fat for 2014 and 2017
Y2014 Animal fat: 209460.54
Y2017 Animal fat: 269617.53


## Question 12.
What is the mean and standard deviation across the whole dataset for the year 2015 to 3 decimal places?

In [4]:
# Solution
# calculating the mean and std to 3 decimal places
mean = round(food_data.Y2015.mean(),3)
std = round(food_data.Y2015.std(),3)

# Printing out the result
print('The mean and standard deviation for 2015 data')
print('=='*22)
print('Mean:', mean)
print('Std:', std)

The mean and standard deviation for 2015 data
Mean: 135.236
Std: 1603.404


## Question 13.
What is the total number and percentage of missing data in 2016 to 2 decimal places?

In [5]:
# solution
# summing up the number of missing data in 2016
missing_data = food_data.Y2016.isnull().sum()

#calculating the percentage of missing data to 2 decimal places
percentage = round(missing_data/len(food_data)*100,2) 

#printing the result
print('Number and Percentage of missing data in 2016')
print('=='*23)
print('Number of missing data:', missing_data)
print(f'Percentage of missing data: {percentage}%')

Number and Percentage of missing data in 2016
Number of missing data: 1535
Percentage of missing data: 2.52%


## Question 14.
Which year had the highest correlation with ‘Element Code’? 

In [6]:
# this code shows how the sellected features are correlated with each other by magnitude (not direction).
food_data[['Element Code','Y2014','Y2015','Y2016','Y2017','Y2018']].corr().abs()

Element Code     Y2014     Y2015     Y2016     Y2017     Y2018
Element Code      1.000000  0.024457  0.023889  0.023444  0.024254  0.024279
Y2014             0.024457  1.000000  0.994647  0.996081  0.995230  0.994872
Y2015             0.023889  0.994647  1.000000  0.995739  0.988048  0.988208
Y2016             0.023444  0.996081  0.995739  1.000000  0.992785  0.992757
Y2017             0.024254  0.995230  0.988048  0.992785  1.000000  0.998103
Y2018             0.024279  0.994872  0.988208  0.992757  0.998103  1.000000

### Observation from above correlation matrix:
The feature `Y2014` has the highest correlation value of `0.024457` with `Element Code`

## Question 15.
What year has the highest sum of Import Quantity?  
**Hint** - Perform a groupby operation on ‘Element’ and use the resulting Dataframe to answer the question

In [7]:
#grouping the data on 'Element', sellect all the "Year" features, sum and store the result in a variable
grouped_data = food_data.groupby('Element')[['Y2014','Y2015','Y2016','Y2017','Y2018']].sum()

#using the loc property to locate the "Import Quantity" index and sorted it in descending order
grouped_data.loc['Import Quantity'].sort_values(ascending=False) 

Y2017    294559.09
Y2018    287997.09
Y2016    286582.78
Y2014    274144.48
Y2015    267018.46
Name: Import Quantity, dtype: float64

### Observation from above result
The result shows that **2017** has the highest sum of Import Quantity with the value **294559.09**

## Question 16. 
What is the total number of the sum of Production in 2014?  
**Hint** - Perform a groupby operation on ‘Element’ and use the resulting Dataframe to answer the question

In [8]:
#using the loc property to access the value of "Production" after summing all the groups
total_2014_production = food_data.groupby('Element')['Y2014'].sum().loc['Production']
print('Total Production in 2014:', total_2014_production)

Total Production in 2014: 1931287.75


## Question 17.
Which of these elements had the highest sum in 2018?  
**Hint** -  Select columns ‘Y2018’ and ‘Element’, Perform a groupby operation on ‘Element’ on the selected dataframe and answer the  question.

In [9]:
# display the top three highest sum after grouping, summing and sorting in descending order
food_data[['Y2018', 'Element']].groupby('Element').sum().sort_values(by='Y2018', ascending=False).head(3)

Y2018
Element                             
Domestic supply quantity  2161192.10
Production                2075072.89
Food                      1303841.28

### Observation from the above result
The result shows that **Domestic supply quantity** had the highest sum in 2018

## Question 18.
Which of these elements had the 3rd lowest sum in 2018?  
**Hint** -  Select columns ‘Y2018’ and ‘Element’, Perform a groupby operation on ‘Element’ on the selected dataframe and answer the  question.

In [10]:
#using the nsmallest() function to return the n number of smallest values after grouping on "Element" and summing
food_data[['Y2018', 'Element']].groupby('Element').sum().nsmallest(3,'Y2018')

Y2018
Element                                         
Tourist consumption                        90.00
Fat supply quantity (g/capita/day)      10258.69
Protein supply quantity (g/capita/day)  11833.56

### Observation from the above result
The result shows that **Protein supply quantity (g/capita/day)** had the 3rd lowest sum *(11833.56)* in 2018

## Question 19.
What is the total Import Quantity in Algeria in 2018?

In [11]:
#filter the dataset for "Algeria" and "Import Quantity", sellect the 'Y2018' feature then sum()
import_quantity = food_data[(food_data.Area == 'Algeria')&(food_data.Element == 'Import Quantity')].Y2018.sum()

#printing result
print('Total Import Quantity in Algeria in 2018')
print('=='*20)
print(f'Import Quantity: {import_quantity}')

Total Import Quantity in Algeria in 2018
Import Quantity: 36238.29


## Question 20.
What is the total number of unique countries in the dataset?

In [12]:
#using the nunique() function to find the total number of unique countries.
print(f"Number of unique countries: {food_data.Area.nunique()}")

Number of unique countries: 49
